libraries/AP_HAL_Linux/Poller.cpp

```diff
@@ -0,0 +1,403 @@
+extern const AP_HAL::HAL& hal;
+
+namespace Linux {
+
+uint32_t Poller::to_epoll_events(Poller::Event ev) {
+    // EPOLLWAKEUP prevents the system from hibernating or suspending when
+    // inside epoll_wait() for this particular event.  It is silently
+    // ignored if the process does not have the CAP_BLOCK_SUSPEND
+    // capability.
+    uint32_t op = EPOLLWAKEUP;
+
+    if (ev & Poller::Event::Read) {
+        op |= EPOLLIN;
+    }
+    if (ev & Poller::Event::Write) {
+        op |= EPOLLOUT;
+    }
+    if (ev & Poller::Event::Error) {
+        op |= EPOLLERR;
+    }
+
+    return op;
+}
+
+bool Poller::register_pollable(Pollable *p, const Poller::Event ev) {
+    if (_epfd < 0) {
+        return false;
+    }
+
+    struct epoll_event epev = {0};
+    epev.events = Poller::to_epoll_events(ev);
+    epev.data.ptr = static_cast<void *>(p);
+
+    return epoll_ctl(_epfd, EPOLL_CTL_ADD, p->get_fd(), &epev) == 0;
+}
+
+void Poller::unregister_pollable(const Pollable *p) {
+    if (_epfd >= 0) {
+        epoll_ctl(_epfd, EPOLL_CTL_DEL, p->get_fd(), NULL);
+    }
+}
+
+int Poller::poll(int timeout_ms) const {
+    const int max_events = 16;
+    epoll_event events[max_events];
+    auto before_wait = AP_HAL::millis();
+    int r = epoll_wait(_epfd, events, max_events, timeout_ms);
+
+    if (r > 0) {
+        const auto delta = AP_HAL::millis() - before_wait;
+        const auto max_time_ms = std::max((timeout_ms - delta) / r, 1U);
+
+        for (int i = 0; i < r; i++) {
+            Pollable *p = static_cast<Pollable *>(events[i].data.ptr);
+
+            if (events[i].events & EPOLLIN) {
+                p->on_can_read(max_time_ms);
+            }
+            if (events[i].events & EPOLLOUT) {
+                p->on_can_write(max_time_ms);
+            }
+            if (events[i].events & EPOLLERR) {
+                p->on_error(max_time_ms);
+            }
+            if (events[i].events & EPOLLHUP) {
+                p->on_hang_up(max_time_ms);
+            }
+        }
+    } else if (r < 0) {
+        if (errno == EINTR) {
+            // A possible nudge from another thread; return 0 to signal
+            // tmout.
+            r = 0;
+        }
+    }
+
+    return r;
+}
+
+Pollable::~Pollable() {
+    auto scheduler = Linux::Scheduler::from(hal.scheduler);
+    scheduler->unregister_pollable(this);
+
+    close(_fd);
+}
+
+bool Pollable::set_blocking(bool setting) {
+    auto curflags = fcntl(_fd, F_GETFL, 0);
+
+    if (curflags < 0) {
+        return false;
+    }
+
+    if (setting) {
+        curflags &= ~O_NONBLOCK;
+    } else {
+        curflags |= O_NONBLOCK;
+    }
+
+    return fcntl(_fd, F_SETFL, curflags) == 0;
+}
+
+void BufferedPollable::on_can_read(int max_time_ms) {
+    if (!_read_sem.take(max_time_ms)) {
+        return;
+    }
+
+    auto space = _read_buffer.space();
+    uint8_t buf[space];
+
+    auto r = read(_fd, buf, space);
+    if (r > 0) {
+        _read_buffer.write(buf, static_cast<uint32_t>(r));
```

Comment:
Good idea. Something similar to `peekiovec` could be used here, so that `readv()` can be used to read directly from the file descriptor to the ring buffer.